app/src/main/java/org/mozilla/fenix/library/history/HistoryAdapter.kt

```diff
@@ -0,0 +1,225 @@
+                }
+                checkbox.setOnCheckedChangeListener(checkListener)
+            }
+        }
+
+        private fun setupMenu() {
+            this.historyMenu = HistoryItemMenu(itemView.context) {
+                when (it) {
+                    is HistoryItemMenu.Item.Delete -> {
+                        item?.apply { actionEmitter.onNext(HistoryAction.Delete.One(this)) }
+                    }
+                }
+            }
+        }
+
+        companion object {
+            const val LAYOUT_ID = R.layout.history_list_item
+        }
+    }
+
+    class HistoryHeaderViewHolder(
+        view: View
+    ) : RecyclerView.ViewHolder(view) {
+        private val title = view.findViewById<TextView>(R.id.history_header_title)
+
+        fun bind(title: String) {
+            this.title.text = title
+        }
+
+        companion object {
+            const val LAYOUT_ID = R.layout.history_header
+        }
+    }
+
+    class HistoryDeleteViewHolder(
+        view: View,
+        private val actionEmitter: Observer<HistoryAction>
+    ) : RecyclerView.ViewHolder(view) {
+        private lateinit var mode: HistoryState.Mode
+
+        private val button = view.findViewById<View>(R.id.delete_history_button).apply {
+            setOnClickListener {
+                val mode = mode
+                if (mode is HistoryState.Mode.Editing && mode.selectedItems.isNotEmpty()) {
+                    actionEmitter.onNext(HistoryAction.Delete.Some(mode.selectedItems))
+                } else {
+                    actionEmitter.onNext(HistoryAction.Delete.All)
+                }
+            }
+        }
+
+        private val text = view.findViewById<TextView>(R.id.delete_history_button_text).apply {
+            val color = ContextCompat.getColor(context, R.color.photonRed60)
+            val drawable = ContextCompat.getDrawable(context, R.drawable.ic_delete)
+            drawable?.setTint(color)
+            this.setCompoundDrawablesWithIntrinsicBounds(drawable, null, null, null)
+        }
+
+        fun bind(mode: HistoryState.Mode) {
+            this.mode = mode
+
+            val text = if (mode is HistoryState.Mode.Editing && mode.selectedItems.isNotEmpty()) {
+                text.context.resources.getString(R.string.history_delete_some, mode.selectedItems.size)
+            } else {
+                text.context.resources.getString(R.string.history_delete_all)
+            }
+
+            button.contentDescription = text
+            this.text.text = text
+        }
+
+        companion object {
+            const val LAYOUT_ID = R.layout.history_delete
+        }
+    }
+
+    private var items: List<HistoryItem> = emptyList()
+    private var mode: HistoryState.Mode = HistoryState.Mode.Normal
+
+    fun updateData(items: List<HistoryItem>, mode: HistoryState.Mode) {
+        this.items = items
+        this.mode = mode
+        notifyDataSetChanged()
```

Comment:
At some point we should get DiffUtil working if anything gets updated asynchronously. I think deleting history would qualify.

Reply:
I agree. I was going to do it inside this PR but it was already starting to get beefy. Once we have the ability to delete we can wire it up and refactor into a DiffUtil.